lib/using_restricted_array.rb

```diff
@@ -1,61 +1,130 @@
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) linear, have to iterate through array until we find it.
+# Space complexity: O(1), constant memory,
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  # raise NotImplementedError
+  length.times do |i|
+    return true if value_to_find == array[i]
+  end
+  false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) linear: since it's unsorted, will have to go through the entire array until the largest value is found.
+# Space complexity: O(1) memory usage does not change for this method, it is constant.
 def find_largest(array, length)
-  raise NotImplementedError
+  #raise NotImplementedError
+  return nil if length.zero?
+
+  largest_value = array[0]
+  i = 1
+  while i < length
+    largest_value = array[i] if array[i] > largest_value
+    i += 1
+  end
+
+  largest_value
+
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:  O(n) linear: since it's unsorted, will have to go through the entire array until the smallest value is found.
+# Space complexity: O(1) memory usage does not change for this method, it is constant.
 def find_smallest(array, length)
-  raise NotImplementedError
+  #raise NotImplementedError
+  return nil if length.zero?
+
+  min_value = array[0]
+  i = 1
+  while i < length
+    min_value = array[i] if array[i] < min_value
+
+    i += 1
+  end
+
+  min_value
+
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)  execute in linear time as compared to the size of the input
+# Space complexity: O(1), since it is in place. New memory is not needed to store the reversed array.
+# Amount of memory used does not change as the size of the input array changes.
 def reverse(array, length)
-  raise NotImplementedError
+  #raise NotImplementedError
+  return nil if length < 1
+
+  first = 0
+  last = length - 1
+
+    while first < last
+      temp = array[first]
+      array[first] = array[last]
+      array[last] = temp
+
+      first += 1
+      last -= 1
+    end
+
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log n), the data set size affects the efficiency of the algorithm in a logarithmic fashion
+# Space complexity: O(1), memory used does not change as the size of the input array changes
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  #raise NotImplementedError
+  return false if length.zero?
+
+  low = 0
+  high = length - 1
+
+  while low < high
```

Comment:
One fix here, for example if the array is of length 1, your code would fail. 

```suggestion
  while low <= high
```